three-cards-divination/src/main/java/com/cyber/mysticism/tarot/controller/ThreeCardsDivinationController.java

```diff
@@ -20,7 +26,32 @@ public ThreeCardsDivinationController(ThreeCardsDivinationService threeCardsDivi
     }
 
     @GetMapping("/three-cards-divination")
-    public Map<String, Card> performReading() throws DivinationException {
-        return threeCardsDivinationService.getReading();
+    public Map<String, Card> performReadingForUser(@RequestHeader(value = "Authorization", required = false) String authHeader)
+            throws DivinationException, ForbiddenAccessException, UnauthorisedAccessException {
+        String[] credentials = authenticateAndParseHeader(authHeader);
+        String username = credentials[0];
+        String email = credentials[1];
+        try {
+            return threeCardsDivinationService.getReadingForUser(username, email);
+        } catch (UserNotFoundException e) {
+            throw new UnauthorisedAccessException("Unauthorized!");
+        }
+    }
+
+    private String[] authenticateAndParseHeader(String authHeader) throws UnauthorisedAccessException, ForbiddenAccessException {
+        if (authHeader == null) {
+            throw new ForbiddenAccessException("Access forbidden!");
+        } else if (!authHeader.isEmpty() && authHeader.toLowerCase().startsWith("basic")) {
+            String base64Credentials = authHeader.substring("Basic".length()).trim();
+            return getCredentialsFromHeaders(base64Credentials);
+        } else {
+            throw new UnauthorisedAccessException("Unauthorized!");
+        }
```

Comment:
**Suggestion:** It is recommended to use `Optional` to handle possible null values in the `authenticateAndParseHeader` method. This can help avoid potential `NullPointerExceptions`. [best practice]
```suggestion
    private Optional<String[]> authenticateAndParseHeader(String authHeader) throws UnauthorisedAccessException, ForbiddenAccessException {
        if (authHeader == null) {
            throw new ForbiddenAccessException("Access forbidden!");
        } else if (!authHeader.isEmpty() && authHeader.toLowerCase().startsWith("basic")) {
            String base64Credentials = authHeader.substring("Basic".length()).trim();
            return Optional.of(getCredentialsFromHeaders(base64Credentials));
        } else {
            throw new UnauthorisedAccessException("Unauthorized!");
        }
    }
```